### ISSUE TITLE
"syntax error: unexpected word (expecting ")")" on Kobo Clara HD

### ISSUE THREAD

**Issue by sagehane** [koreader/koreader#8076]
* KOReader version: v2021.07
* Device: Kobo Clara HD (FW 4.26.16704)

#### Issue
KOReader crashes when started.
#### Steps to reproduce
KOReader was installed by the semi-automated install method described [here](https://github.com/koreader/koreader/wiki/Installation-on-Kobo-devices#semi-automated-installation-method).
Going from the NickelMenu -> KOReader causes a crash.
<details>
  <summary>crash.log</summary>

```
Original fb rotation is set @ 3
Original fb bitdepth is set @ 32bpp
Switching fb bitdepth to 8bpp & rotation to Portrait
[FBInk] Detected a Kobo Clara HD (376 => Nova @ Mark 7)
Variable fb info: 1072x1448 (1088x1536), 32bpp @ rotation: 3 (Counter Clockwise, 270°)
Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 6782976 bytes & line length: 4352 bytes

Requested canonical rota 0 translates to 3 for this device

Current rotation is already 3!

Switching fb to 8bpp @ rotation 3 . . .
Setting bitdepth to 8bpp
Setting grayscale to 1
Setting rotate to 3 (Counter Clockwise, 270°)
Bitdepth is now 8bpp (grayscale: 1) @ rotate: 3 (Counter Clockwise, 270°)

Variable fb info: 1072x1448 (1088x3072), 8bpp @ rotation: 3 (Counter Clockwise, 270°)
Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 6782976 bytes & line length: 1088 bytes
./reader.lua: line 2: syntax error: unexpected word (expecting ")")
!!!!
Uh oh, something went awry... (Crash n°1: 08/14/21 @ 20:11:16)
Running FW 4.26.16704 on Linux 4.1.15-00089-ga2737fc02713 (#34 SMP PREEMPT Fri Dec 27 17:47:13 CST 2019)
Attempting to restart KOReader . . .
!!!!
./reader.lua: line 2: syntax error: unexpected word (expecting ")")
!!!!
Uh oh, something went awry... (Crash n°2: 08/14/21 @ 20:11:31)
Running FW 4.26.16704 on Linux 4.1.15-00089-ga2737fc02713 (#34 SMP PREEMPT Fri Dec 27 17:47:13 CST 2019)
Attempting to restart KOReader . . .
!!!!
./reader.lua: line 2: syntax error: unexpected word (expecting ")")
!!!!
Uh oh, something went awry... (Crash n°3: 08/14/21 @ 20:11:31)
Running FW 4.26.16704 on Linux 4.1.15-00089-ga2737fc02713 (#34 SMP PREEMPT Fri Dec 27 17:47:13 CST 2019)
Attempting to restart KOReader . . .
!!!!
./reader.lua: line 2: syntax error: unexpected word (expecting ")")
!!!!
Uh oh, something went awry... (Crash n°4: 08/14/21 @ 20:11:31)
Running FW 4.26.16704 on Linux 4.1.15-00089-ga2737fc02713 (#34 SMP PREEMPT Fri Dec 27 17:47:13 CST 2019)
Attempting to restart KOReader . . .
!!!!
./reader.lua: line 2: syntax error: unexpected word (expecting ")")
!!!!
Uh oh, something went awry... (Crash n°5: 08/14/21 @ 20:11:32)
Running FW 4.26.16704 on Linux 4.1.15-00089-ga2737fc02713 (#34 SMP PREEMPT Fri Dec 27 17:47:13 CST 2019)
Too many consecutive crashes, aborting . . .
!!!! ! !!!!
Restoring original fb bitdepth @ 32bpp & rotation @ 3
[FBInk] Detected a Kobo Clara HD (376 => Nova @ Mark 7)
Variable fb info: 1072x1448 (1088x3072), 8bpp @ rotation: 3 (Counter Clockwise, 270°)
Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 6782976 bytes & line length: 1088 bytes

Current rotation is already 3!

Switching fb to 32bpp @ rotation 3 . . .
Setting bitdepth to 32bpp
Setting grayscale to 0
Setting rotate to 3 (Counter Clockwise, 270°)
Bitdepth is now 32bpp (grayscale: 0) @ rotate: 3 (Counter Clockwise, 270°)

Variable fb info: 1072x1448 (1088x1536), 32bpp @ rotation: 3 (Counter Clockwise, 270°)
Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 6782976 bytes & line length: 4352 bytes
```
</details>

#### Things to note
I was initially using KOReader just fine on versions 2021.04 to 2021.07 on: 
```
FW 4.21.15023 on Linux 4.1.15-00089-ga2737fc02713 (#34 SMP PREEMPT Fri Dec 27 17:47:13 CST 2019)
```
I decided to update my firmware, and that broke NickelMenu with this error:
```
kfmon: error checking '/tmp/kfmon-ipc.ctl':stat: No such file or directory (src/generator_c.c:93)
```
I then redid the semi-automated install step again (using the OCP-KOReader-v2021.07-6.zip from [here](https://storage.gra.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/kfmon-pub/kfmon.html) as opposed to the v2021.04 that I used the first time around), and then came across the error described on the issue.

**Comment by Frenzie** [comment]
There's no syntax error unless your file was damaged during extraction somehow. Pinging @NiLuJe

**Comment by NiLuJe** [comment]
The most logical explanation I can think of would be severe corruption somewhere along the line :?.

(i.e., this sort of stuff happens when you feed the dynamic loader broken binaries, or binaries from a wrong architecture...).

**Comment by NiLuJe** [comment]
I would *also* reinstall the latest FW manually first, just in case something went awry there, too...

----

Other than that, is there anything else custom installed on this device?

**Comment by sagehane** [comment]
>There's no syntax error unless your file was damaged during extraction somehow.

I suspected that the `reader.lua` might have been corrupted, so I ran a simple sha256sum check before opening the issue and nothing seems to be wrong with the file itself. Are there any other files worth checking the integrity of?

>I would also reinstall the latest FW manually first, just in case something went awry there, too...

Also, I'm quite a noob when it comes to where to find resources for these kind of stuff. How do I do this?

>EDIT: Use a fresh download for everything, in case the corruption is on your computer's/network's end and not the device's.

I checked for the hashsum of the script I ran (I tried downloading a fresh copy and the hashsum was the same as the one I downloaded on April) and the OCP KOReader file. So the only possible point of failure is during the extraction/installation process. I'll try it again, sure.

>Other than that, is there anything else custom installed on this device?

Absolutely nothing. The only tweak I made was following the instructions of [this blog](https://yingtongli.me/blog/2018/07/30/kobo-rego.html) to bypass the need for registering.

**Comment by Frenzie** [comment]
> I suspected that the reader.lua might have been corrupted, so I ran a simple sha256sum check before opening the issue and nothing seems to be wrong with the file itself. Are there any other files worth checking the integrity of?

All of them, ;-) but if it's not even getting past line 2 of the very first script file it's either that script file and/or the luajit binary. There are no other options at that point yet. Which doesn't exclude anything else from also being damaged of course.

**Comment by sagehane** [comment]
Sounds like an awfully specific corruption on the luajit binary, lol

**Comment by NiLuJe** [comment]
Given the output, the corruption is likely (at least) in the luajit binary (or you're extremely unlucky, and the reader.lua hashbang is still a hashbang but pointing at a garbage binary ^^).

**Comment by sagehane** [comment]
Welp, I tried running the install script and it worked fine this time (???). Sorry to waste you two's time. I really should've given it a try once more. Much thanks for the very fast response.

**Comment by NiLuJe** [comment]
Just for curiosity's sake:

* Do you remember if you unplugged the device *without* a proper Eject the first time around?
* Are you on an M1 mac?
* Were you using an USB 3.x port?
* If you're on Windows, is it setup for safe ejects or not (I, err, can't remember the exact wording and where the damn setting is set, though).

TL;DR:

* FW 4.28 has reliability fixes for USB issues
* *Always* eject first, don't trust your OS. It's usually explicit on macOS & Linux, but on Windows, prefer right clicking the drive to find the actual "Eject" entry to the systray USB thingy.

**Comment by sagehane** [comment]
- I'm on NixOS Unstable. I generally use Linux. (I should've stated this upfront, tbh)
- I ran `sync`, `umount`, and then `eject` for like the 8+ times I inserted and uninserted the micro SD today.
- I was using an USB 3.0 port, with a micro SD-to-USB-A adapter.

Also, I would appreciate it if you could redirect me to a place where I can check for firmware updates, release notes, and how to install them?

**Comment by NiLuJe** [comment]
Whelp, let's chalk that one up to Gremlins, then ;).

@pgaskin runs a neat little webservice for that: https://pgaskin.net/KoboStuff/kobofirmware.html ;).

**Comment by NiLuJe** [comment]
>     * I ran `sync`, `umount`, and then `eject` for like the 8+ times I inserted and uninserted the micro SD today.

Err, are you actually pulling out the *internal* SD card? o_O.
> 
>     * I was using an USB 3.0 port, with a micro SD-to-USB-A adapter. (Edit: the adapter itself seems to support the 3.0 interface? It's blue on the tip, but idk how adapters work)

I'd shy away from all manners of hubs, extenders & adatpers as far as Kobos & Kindles are concerned ;).

**Comment by sagehane** [comment]
>Err, are you actually pulling out the internal SD card? o_O.

Yes... uhhh, was I not supposed to do that? Okay... I was just told from another guy who owns the device that that's what he did. I wasn't aware of any other way... Maybe I should've done more reading...

I mean, SSH is a thing? But that seems to only work _after_ bootstrapping KOReader, and the Dropbear package doesn't seem to support an ed25519 key yet[0], welp. (I presume KOReader is based on Debian Buster packages atm?)

So I guess I'm supposed to use the micro USB normally?

[0] https://github.com/mkj/dropbear/pull/75

**Comment by NiLuJe** [comment]
Err, yeah, Joe Random is indeed supposed to just plug it in and enjoy USBMS access to the user side storage partition ;).

If you actually need rootfs access, I have a bunch of tools with recent builds [over here](http://www.mobileread.com/forums/showthread.php?t=225030) ;).

**Comment by NiLuJe** [comment]
> I mean, SSH is a thing? But that seems to only work _after_ bootstrapping KOReader, and the Dropbear package doesn't seem to support an ed25519 key yet[0], welp. (I presume KOReader is based on Debian Buster packages atm?)
> 

No, it's just that this requires an insane amount of custom patching to make it behave in KOReader's context, and I refuse to be saddled with maintaining it, because KOReader is not the right place to do it ;). The lack of elliptical curves is just the bitrot setting in ;).

**Comment by sagehane** [comment]
>Err, yeah, Joe Random is indeed supposed to just plug it in and enjoy USBMS access to the user side storage partition ;).

I find it hard to find words that can describe how dumb I feel about myself right now. How did I not even _think_ about trying that?

>No, it's just that this requires an insane amount of custom patching to make it behave in KOReader's context, and I refuse to be saddled with maintaining that plugin, because KOReader is not the right place to do it ;).

Oh well. I was really confused when SSH didn't work for me (especially considering I didn't know much about it). Is there at least a way to give a visual indicator that it's not supported?

Either way, thanks for all the help. Appreciate it a lot.
